Use an efficient representation for merged components of operations

codrut3 opened the pull request:
Currently [`merge_operations`](https://github.com/quantumlib/Cirq/blob/a89f5b9f0875dcae419efc13790af078075fcb6b/cirq-core/cirq/transformers/transformer_primitives.py#L357) represents merged components as a `CircuitOperation`. This means in a merge of `n` operations, `n-1` `CircuitOperations` are created, with a complexity of `O(n^2)`.
I use a disjont-set data structure to reduce the complexity to `O(n)` for `merge_k_qubit_unitaries_to_circuit_op`. `merge_operations` itself can't be improved because it uses a `merge_func` that requires creating a `CircuitOperation` at every step.

codrut3 commented:
I didn't change the algorithm in merge operations, but switched from CircuitOperation to Component for intermediate merges.
I wrote an explanation for the algorithm [here](https://docs.google.com/document/d/179nk8NMo-VtUgExSm73oU2oLLp7oGfUuVfZ1-RWiYi8/edit?usp=sharing).

I found this inefficiency while investigating why `merge_single_qubit_gates_to_phased_x_and_z` in the Shannon decomposition is slow. Right now this method is not called anymore by the Shannon decomposition, I replaced it with a custom version that is faster.

Still this change improves transpilation. On my laptop, I measured an average improvement of 5,09% from 86.89 seconds to 82,46 seconds for 7 qubits.

I also measured the improvement, if `merge_single_qubit_gates_to_phased_x_and_z` would still be used in the Shannon decomposition. In this case it is 20.7% from 40.88 seconds to 32.40 seconds for 8 qubits.

This is the script I used to measure the transpiler improvement:
```python
"""Tests performance of quantum Shannon decomposition.
"""

import sys
import time

from scipy.stats import unitary_group

import cirq


if not sys.warnoptions:
    import warnings
    warnings.simplefilter("ignore")


def main():
    measurements = []
    n_qubits = 7
    for i in range(10):
        print(f'Iteration {i}')
        U = unitary_group.rvs(2**n_qubits)
        qubits = [cirq.NamedQubit(f'q{i}') for i in range(n_qubits)]
        start_time = time.time()
        circuit = cirq.Circuit(cirq.quantum_shannon_decomposition(qubits, U))
        elapsed_time = time.time() - start_time
        print(f'Time taken: {elapsed_time:.2f} seconds')

        # Run cirq transpiler
        start_time = time.time()
        circuit = cirq.optimize_for_target_gateset(circuit, gateset=cirq.CZTargetGateset(preserve_moment_structure=False), max_num_passes=None)
        elapsed_time = time.time() - start_time
        measurements.append(elapsed_time)
        print(f'Cirq depth for {n_qubits} qubits (with transpiler): {len(circuit)} ({elapsed_time:.2f} sec)')

    print(f'Average time taken for transpiling: {sum(measurements) / len(measurements):.2f} seconds')


if __name__ == '__main__':
    main()
```

babacry reviewed (commented):
Overall, I think the PR greatly improved the performance and the readability!

The unit tests look solid and confirm the merge behavior is working as intended. I've focused my review on the overall design in this iteration.

For performance, IIUC, given m connected ops,
* before: construct CircuitOperation m-1 times.
* after: construct CircuitOperation 1 time + extra on-average-constant disjoint-set union and find time.
The bottleneck is the CircuitOperation construction time in the previous implementation. Correct me if I am wrong.

Since this is a big design change, we should be careful here for the compatibility and extensibility. I've put a couple of comments for discussion.

codrut3 commented on cirq-core/cirq/transformers/connected_component.py:
```diff
@@ -0,0 +1,278 @@
+        if not x.is_mergeable or not y.is_mergeable or not x.can_merge(x.ops, y.ops):
+            return None
+
+        root = cast(ComponentWithOps, super(ComponentWithOps, x).merge(y, merge_left))
```
merge_left decides if x' or y's moment will be copied to the new representative. This saves the user the trouble of copying the moment over to the new representative. It's equivalent to y.merge(x, !merge_left).

codrut3 commented on cirq-core/cirq/transformers/connected_component.py:
```diff
@@ -0,0 +1,278 @@
+        self.ops = [op]
+        self.can_merge = can_merge
+
+    def merge(self, c: Component, merge_left=True) -> Component | None:
```
Done.

codrut3 commented on cirq-core/cirq/transformers/connected_component.py:
```diff
@@ -0,0 +1,278 @@
+    is_mergeable: bool
+
+    # Circuit moment containing the component
+    moment: int
```
Done.

babacry commented on cirq-core/cirq/transformers/connected_component.py:
```diff
@@ -0,0 +1,278 @@
+        if not x.is_mergeable or not y.is_mergeable:
+            return None
+
+        new_op = x.merge_func(x.circuit_op, y.circuit_op)
```
I don't know how either. Could potentially be useful as there might be Component specific merge directions that are not uniform.

babacry commented on cirq-core/cirq/transformers/transformer_primitives.py:
```diff
@@ -304,54 +312,224 @@ class _MergedCircuit:
+    circuit: CIRCUIT_TYPE,
+    factory: ComponentFactory,
+    *,
+    merged_circuit_op_tag: str = "Merged connected component",
```
Is it a good practice to put space in a tag? @NoureldinYosri  I am not sure, though existing code before this pr have some mixed style of tags.

maybe "merged_connected_component"? or "_merged_connected_component" (start with _ to separate `a default tag named by cirq pkg` with `user specified tag`)?

codrut3 replied:
I copied the tag value from line 621 in merge_operations_to_circuit_op. I think the two should be in sync.

NoureldinYosri reviewed (commented):
@codrut3  sorry about the late reply but I was wondering if instead of manually implementing union-find we can do the required the functionality using https://networkx.org/documentation/stable/_modules/networkx/utils/union_find.html

this will split the logic of merging moments into two parts where one of them where we need to maintain only one of them

```py3
   def merge(self, a, b):
          self._merge_moments(a, b)
          self.union_find.union(a, b)
```

is it possible to do that or will it break the logic?

codrut3 commented on cirq-core/cirq/transformers/transformer_primitives.py:
```diff
@@ -422,63 +596,16 @@ def apply_merge_func(op1: ops.Operation, op2: ops.Operation) -> ops.Operation |
-                    merged_circuit.add_op_to_moment(moment_idx, op)
-                continue
+    def is_mergeable(op: cirq.Operation):
+        del op
```
Done. I used _ instead.

codrut3 commented on cirq-core/cirq/transformers/transformer_primitives.py:
```diff
@@ -282,17 +290,17 @@ def map_operations_and_unroll(
+            containing measurement operations with the same key.
+        ckey_indexes: Mapping from measurement keys to (sorted) list of component moments
+            containing classically controlled operations controlled on the same key.
+        components_by_index: List of circuit moments containing components. We use a dictionary
```
The name is correct, but confusing. The data structure is:
list of dictionary {key: component, value: 0}, where the position in the list is given by the moment id

So the dictionaries are in a list, with the index the moment id. For a given moment, a dictionary is used to preserve the insertion order of the components in the moment. The value has no meaning, it's only there to take advantage of the dict order preserving feature.

I added more details in the docstring, please have a look!

babacry replied:
Ah, I see. It's a bit tricky

```
components_by_index: List of components indexed by moment id.
            For a moment id, we use a dictionary to keep track of the
            components in the moment. The dictionary instead of a set is used to preserve insertion order and
            the dictionary's values are intentionally unused.
```

codrut3 replied:
Thanks, I updated the description.

codrut3 commented:
> @codrut3 sorry about the late reply but I was wondering if instead of manually implementing union-find we can do the required the functionality using https://networkx.org/documentation/stable/_modules/networkx/utils/union_find.html
> 
> this will split the logic of merging moments into two parts where one of them where we need to maintain only one of them
> 
> ```python
>    def merge(self, a, b):
>           self._merge_moments(a, b)
>           self.union_find.union(a, b)
> ```
> 
> is it possible to do that or will it break the logic?

I think it works to use a separate data structure to do the union-find. This structure would have to be initialized by the user and passed as an argument to the component factory.

However, I would prefer to use [scipy.DisjointSet](https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.DisjointSet.html) because it allows me to add new singletons on the fly, while https://networkx.org/documentation/stable/_modules/networkx/utils/union_find.html requires all elements to be known from the beginning.

If you are ok with adding a dependency on the scipy package, I will rewrite the code. Do I need to add the package version as a requirement somewhere?

Let me know how to proceed! @NoureldinYosri

codrut3 commented on cirq-core/cirq/transformers/transformer_primitives.py:
```diff
@@ -304,54 +312,224 @@ class _MergedCircuit:
+            merged_circuit_op_tag: tag to use for CircuitOperations
+
+        Returns:
+            the circuit with merged components as a CircuitOperation
```
Done.

codrut3 commented:
> @codrut3 that sounds good, Cirq already depends on scipy
> 
> https://github.com/quantumlib/Cirq/blob/fe0dc2187ca3269c178526e8ba41083fa1a467c9/cirq-core/requirements.txt#L10
> 
> sorry for the extra work

No worries! I changed the implementation to use `scipy.DisjointSet`.

I took this opportunity to rewrite the connected component classes. I moved the `merge()` and `find()` methods to the factory class, which I renamed as `ComponentSet`. This solved the issue that `can_merge` and `merge_func` were defined at the component level, and could potentially differ for the two components at merge time. Now they are defined at the level of `ComponentSet` which ensures all created components have the same callable (as well as are part of the same `scipy.DisjointSet`).

pavoljuhas commented on cirq-core/cirq/transformers/_connected_component.py:
```diff
@@ -0,0 +1,240 @@
+    # Union of all control keys in the component
+    ckeys: frozenset[cirq.MeasurementKey]
+    # Initial operation in the component
+    op: cirq.Operation
```
I see - if Components are mutable and should be used in sets and dictionaries, it is indeed better to leave as is and rely on object identity.

pavoljuhas commented on cirq-core/cirq/transformers/transformer_primitives.py:
```diff
@@ -304,54 +342,203 @@ class _MergedCircuit:
         default_factory=lambda: defaultdict(lambda: [-1])
     )
-    ops_by_index: list[dict[cirq.Operation, int]] = dataclasses.field(default_factory=list)
+    components_by_index: list[dict[Component, int]] = dataclasses.field(default_factory=list)
```
Nit - if dictionary values are unused, it is perhaps better to declare it as `dict[Component, NoneType]` and use `None` instead of `0` to add to dictionary.

pavoljuhas reviewed (approved):
LGTM after addressing some minor issues - mainly renaming the new module `connected_component` to `_connected_component` to mark it as for internal use only.

Rather than writing a lot of comments I pushed the suggestions directly to the PR, these should be easy to review commit-by-commit.

Please let me know if this looks OK to you, LGTM on my side (with one tiny inline suggestion).

codrut3 commented on cirq-core/cirq/transformers/transformer_primitives.py:
```diff
@@ -49,6 +58,29 @@ def _create_target_circuit_type(ops: ops.OP_TREE, target_circuit: CIRCUIT_TYPE)
+
+    Optimized for the majority case when the value is the last in the list.
+    """
+    indices.reverse()
```
I made some changes in my commit, please have a look!

pavoljuhas reviewed (approved):
Good to go on my side.

pavoljuhas commented:
One last tweak in 7637ddea6e5d4eaa6bc44671adb32e84f5d8f3ca - typing_extensions is not in the cirq-core requirements anymore so I removed its import.

codrut3 commented:
Nice, looks good to me, thank you a lot for the changes!
